pkg/controller/replicaset/replica_set_utils.go

```diff
@@ -85,3 +84,88 @@ func (o overlappingReplicaSets) Less(i, j int) bool {
 	}
 	return o[i].CreationTimestamp.Before(o[j].CreationTimestamp)
 }
+
+func calculateStatus(rs extensions.ReplicaSet, filteredPods []*api.Pod, manageReplicasErr error) extensions.ReplicaSetStatus {
+	newStatus := rs.Status
+	// Count the number of pods that have labels matching the labels of the pod
+	// template of the replica set, the matching pods may have more
+	// labels than are in the template. Because the label of podTemplateSpec is
+	// a superset of the selector of the replica set, so the possible
+	// matching pods must be part of the filteredPods.
+	fullyLabeledReplicasCount := 0
+	readyReplicasCount := 0
+	availableReplicasCount := 0
+	templateLabel := labels.Set(rs.Spec.Template.Labels).AsSelectorPreValidated()
+	for _, pod := range filteredPods {
+		if templateLabel.Matches(labels.Set(pod.Labels)) {
+			fullyLabeledReplicasCount++
+		}
+		if api.IsPodReady(pod) {
+			readyReplicasCount++
+			if api.IsPodAvailable(pod, rs.Spec.MinReadySeconds, unversioned.Now()) {
+				availableReplicasCount++
+			}
+		}
+	}
+
+	if manageReplicasErr != nil {
+		var reason string
+		if diff := len(filteredPods) - int(rs.Spec.Replicas); diff < 0 {
+			reason = "FailedCreate"
+		} else if diff > 0 {
+			reason = "FailedDelete"
+		}
+		cond := NewReplicaSetCondition(extensions.ReplicaSetReplicaFailure, api.ConditionTrue, reason, manageReplicasErr.Error())
+		SetCondition(&newStatus, cond)
+	} else {
+		RemoveCondition(&newStatus, extensions.ReplicaSetReplicaFailure)
+	}
+
+	newStatus.Replicas = int32(len(filteredPods))
+	newStatus.FullyLabeledReplicas = int32(fullyLabeledReplicasCount)
+	newStatus.ReadyReplicas = int32(readyReplicasCount)
+	newStatus.AvailableReplicas = int32(availableReplicasCount)
+	return newStatus
+}
+
+func NewReplicaSetCondition(condType extensions.ReplicaSetConditionType, status api.ConditionStatus, reason, msg string) extensions.ReplicaSetCondition {
+	return extensions.ReplicaSetCondition{
+		Type:               condType,
+		Status:             status,
+		LastTransitionTime: unversioned.Now(),
+		LastProbeTime:      unversioned.Now(),
+		Reason:             reason,
+		Message:            msg,
+	}
+}
+
+func GetCondition(status extensions.ReplicaSetStatus, condType extensions.ReplicaSetConditionType) *extensions.ReplicaSetCondition {
+	for i := range status.Conditions {
+		c := status.Conditions[i]
+		if c.Type == condType {
+			return &c
+		}
+	}
+	return nil
+}
+
+func SetCondition(status *extensions.ReplicaSetStatus, cond extensions.ReplicaSetCondition) {
+	newConditions := filterOutCondition(status.Conditions, cond.Type)
+	status.Conditions = append(newConditions, cond)
```

Comment:
if the condition already is set, will this just remove it and add it back again?